File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int n = friends.length;
        
        Map<String, Integer> friendIdx = new HashMap<>();
        for (int i = 0; i < friends.length; i++) {
            friendIdx.put(friends[i], i);
        }

        int[][] giftMatrix = new int[n][n];
        for (String gift : gifts) {
            String[] parts = gift.split(" ");
            String sender = parts[0];
            String receiver = parts[1];
            int from = friendIdx.get(sender);
            int to = friendIdx.get(receiver);
            
            giftMatrix[from][to]++;
        }
        
        int[] giftScore = new int[n];
        for (int i = 0; i < n; i++) {
            int given = 0;
            int received = 0;
            
            for (int j = 0; j < n; j++) {
                given += giftMatrix[i][j];
                received += giftMatrix[j][i];
            }
            
            giftScore[i] = given - received;
        }
        
        int[] receiveCnt = new int[n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == j) continue;
                
                int given = giftMatrix[i][j];
                int received = giftMatrix[j][i];
                
                if ((given > received) ||
                    (given == received && giftScore[i] > giftScore[j])) {
                    receiveCnt[i]++;
                }
            }
        }
        
        int answer = 0;
        for (int cnt : receiveCnt) {
            answer = Math.max(answer, cnt);
        }
        
        return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.kt
class Solution {
    fun solution(friends: Array<String>, gifts: Array<String>): Int {
        val n = friends.size
        
        val friendIdx = friends.withIndex().associate { it.value to it.index }
        
        val giftMatrix = Array(n) { IntArray(n) }
        for (gift in gifts) {
            val (sender, receiver) = gift.split(" ")
            val from = friendIdx[sender]!!
            val to = friendIdx[receiver]!!
            
            giftMatrix[from][to]++
        }
        
        val giftScore = IntArray(n)
        for (i in 0 until n) {
            var given = 0
            var received = 0
            
            for (j in 0 until n) {
                given += giftMatrix[i][j]
                received += giftMatrix[j][i]
            }
            
            giftScore[i] = given - received
        }
        
        val receiveCnt = IntArray(n)
        for (i in 0 until n) {
            for (j in 0 until n) {
                if (i == j) continue
                
                val given = giftMatrix[i][j]
                val received = giftMatrix[j][i]
                
                if ((given > received) ||
                    (given == received && giftScore[i] > giftScore[j])) {
                    receiveCnt[i]++
                }
            }
        }
        
        return receiveCnt.maxOrNull() ?: 0
    }
}